src/java.base/share/classes/jdk/internal/lang/stable/CachedIntFunction.java

```diff
@@ -0,0 +1,42 @@
+package jdk.internal.lang.stable;
+
+import jdk.internal.vm.annotation.ForceInline;
+
+import java.util.List;
+import java.util.function.IntFunction;
+
+// Note: It would be possible to just use `LazyList::get` instead of this
+// class but explicitly providing a class like this provides better
+// debug capability, exception handling, and may provide better performance.
+public record CachedIntFunction<R>(List<StableValueImpl<R>> stables,
+                                   IntFunction<? extends R> original) implements IntFunction<R> {
+    @ForceInline
+    @Override
+    public R apply(int value) {
+        final StableValueImpl<R> stable;
+        try {
+            // Todo: Will the exception handling here impair performance?
```

Comment:
Can we add a custom IAE formatter, like the String/ArrayIOOBE formatters in `jdk.internal.util.Preconditions` and use it to check bounds before we call `get`?

Reply:
There is a separate effort going on with this.